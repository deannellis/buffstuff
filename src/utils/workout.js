import moment from 'moment';

export const getWorkouts = ({ a = 0, b = 0 }) => {
    let workouts = []
    if(a == 0) {
        workouts.push('overhead', 'chinup')
    } else {
        workouts.push('bench', 'row')
    }
    if(b == 0 || b == 2) {
        workouts.push('squat');
    } else {
        workouts.push('deadlift');
    }
    return workouts;
}

export const getDisplayName = name => {
    switch(name) {
        case 'bench':
            return 'Bench Press';
        case 'row':
            return 'Row';
        case 'squat':
            return 'Squat';
        case 'deadlift':
            return 'Deadlift';
        case 'overhead':
            return 'Overhead Press';
        case 'chinup':
            return 'Chin-Ups';
        default:
            return name;
    }
};

export const getWarmupWeights = workWeight => {
    const warmup25 = workWeight * 0.25 > 44 ? workWeight * 0.25 : 44;
    const warmup50 = workWeight * 0.50 > 44 ? workWeight * 0.50 : 44;
    const warmup75 = workWeight * 0.75 > 44 ? workWeight * 0.75 : 44;
    return [ warmup25, warmup50, warmup75 ];
}

export const getWeightDistribution = weight => {
    if(weight === 44) {
        return [ 0, 44, 0]
    }
    const oneSide = (weight - 44) / 2;
    return [ oneSide, 44, oneSide ];
}

export const getMonthWorkouts = workouts => {
    let dates = {};
    workouts.forEach(workout => {
        let month = moment(workout.created).format("MMMM");
        let day = parseInt(moment(workout.created).format("D"));
        if(dates.hasOwnProperty(month)) {
            dates[month].push(day);
        } else {
            dates[month] = [ day ];
        }
    });
    // console.log(dates);
    return dates;
}

export const getWorkoutDays = (start, daysInMonth) => {
    let workoutDays = [];
    let checkEven = false;
    let weekday = start + 1;
    for(let d = 1; d <= daysInMonth; d++) {
        if(checkEven) {
            if(d%2 == 0) workoutDays.push(d)
        } else {
            if(d%2 !== 0) workoutDays.push(d)
        }
        if(weekday == 7){
            weekday = 0;
            checkEven = !checkEven;
        };
        weekday ++;
    }
    return workoutDays;
};
