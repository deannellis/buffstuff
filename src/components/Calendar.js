import React, { Component } from 'react';
import moment from 'moment';
import { getMonthWorkouts, getWorkoutDays, getWorkoutIds } from '../utils/workout';
import { LeftArrowIcon, RightArrowIcon } from '../utils/icons';

class Calendar extends Component {
    constructor(props) {
        super(props);
        this.state = {
            dateObject: moment(),
        }
        this.getFirstDayOfMonth = this.getFirstDayOfMonth.bind(this);
    }

    getFirstDayOfMonth = () => {
        let { dateObject } = this.state;
        return moment(dateObject).startOf('month').format('d');
    };
    getDaysInMonth = () => {
        return this.state.dateObject.daysInMonth();
    };
    getCurrentDay = () => {
        return this.state.dateObject.format("D");
    };
    getMonth = () => {
        return this.state.dateObject.format("MMMM");
    };
    getYear = () => {
        return this.state.dateObject.format("YYYY");
    };
    onPrevMonth = () => {
        this.setState({
            dateObject: this.state.dateObject.subtract(1, 'month')
        });
    };
    onNextMonth = () => {
        this.setState({
            dateObject: this.state.dateObject.add(1, 'month')
        });
    };
    handleDayClick = id => {
        this.props.onClickWorkoutDate(id);
    };


    render() { 
        const weekdaysAbbreviated = ['S', 'M', 'T', 'W', 'T', 'F', 'S'];
        const workouts = getMonthWorkouts(this.props.workouts, this.getYear());
        const daysWorkedOut = workouts[this.state.dateObject.format("MMMM")] ? workouts[this.state.dateObject.format("MMMM")] : [];
        const workoutDays = getWorkoutDays(parseInt(this.getFirstDayOfMonth()), this.getDaysInMonth());
        const workoutIds = getWorkoutIds(this.props.workouts, this.getMonth());
        let emptyCells = [];
        let daysInMonth = [];
        for(let i = 0; i < this.getFirstDayOfMonth(); i++) {
            emptyCells.push(
                <div className="calendar__day--empty" key={i}></div>
            );
        }
        for(let d = 1; d <= this.getDaysInMonth(); d++) {
            let workedOut = false;
            let isToday = false;
            let workOutDay = false;
            if(moment().format("MMMM") == this.getMonth()) {
                if(d == this.getCurrentDay()) {
                    isToday = true;
                }
            }
            if(daysWorkedOut.includes(d)) workedOut = true;
            if(workoutDays.includes(d)) workOutDay = true;
            daysInMonth.push(
                <div 
                    className={`calendar__day
                        ${isToday ? 'calendar__day--today' : ''}
                        ${workedOut ? 'calendar__day--worked-out' : ''}
                        ${workOutDay ? 'calendar__day--work-out' : ''}
                    `} 
                    key={d}
                    onClick={workedOut ? () => {this.handleDayClick(workoutIds[d])} : () => {}}
                >{d}</div>
            );
        }
        return (
            <div className="calendar">
                <div className="calendar__prev-month" onClick={e => {this.onPrevMonth()}}>
                    <LeftArrowIcon size={24} />
                </div>
                <div className="calendar__month-year">{this.getMonth() + ' ' + this.getYear()}</div>
                <div className="calendar__next-month" onClick={e => {this.onNextMonth()}}>
                    <RightArrowIcon size={24} />
                </div>
                {weekdaysAbbreviated.map((day, i) => (
                    <div className="calendar__weekday-short-name" key={i} >{day}</div>
                ))}
                {emptyCells}
                {daysInMonth}
            </div>
        );
    }
}

Calendar.defaultProps = {
    workouts: [],
    onClickWorkoutDate: () => {console.log('workout date clicked')},
}
 
export default Calendar;
