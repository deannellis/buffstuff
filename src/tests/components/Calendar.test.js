import React from 'react';
import { shallow } from 'enzyme';

import Calendar from '../../components/Calendar';
import { workouts } from '../fixtures/workout';

test('should render Calendar without props', () => {
    const wrapper = shallow(<Calendar />);
    expect(wrapper).toMatchSnapshot();
});

test('should render Calendar with props', () => {
    const wrapper = shallow(<Calendar workouts={workouts} />);
    expect(wrapper).toMatchSnapshot();
});