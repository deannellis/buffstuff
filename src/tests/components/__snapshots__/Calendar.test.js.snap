// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Calendar with props 1`] = `
<div
  className="calendar"
>
  <div
    className="calendar__prev-month"
    onClick={[Function]}
  >
    <LeftArrowIcon
      size={24}
    />
  </div>
  <div
    className="calendar__month-year"
  >
    December 1969
  </div>
  <div
    className="calendar__next-month"
    onClick={[Function]}
  >
    <RightArrowIcon
      size={24}
    />
  </div>
  <div
    className="calendar__weekday-short-name"
    key="0"
  >
    S
  </div>
  <div
    className="calendar__weekday-short-name"
    key="1"
  >
    M
  </div>
  <div
    className="calendar__weekday-short-name"
    key="2"
  >
    T
  </div>
  <div
    className="calendar__weekday-short-name"
    key="3"
  >
    W
  </div>
  <div
    className="calendar__weekday-short-name"
    key="4"
  >
    T
  </div>
  <div
    className="calendar__weekday-short-name"
    key="5"
  >
    F
  </div>
  <div
    className="calendar__weekday-short-name"
    key="6"
  >
    S
  </div>
  <div
    className="calendar__day--empty"
    key="0"
  />
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="1"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="2"
    onClick={[Function]}
  >
    2
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="3"
    onClick={[Function]}
  >
    3
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="4"
    onClick={[Function]}
  >
    4
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="5"
    onClick={[Function]}
  >
    5
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="6"
    onClick={[Function]}
  >
    6
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="7"
    onClick={[Function]}
  >
    7
  </div>
  <div
    className="calendar__day
                        
                        calendar__day--worked-out
                        calendar__day--work-out
                    "
    key="8"
    onClick={[Function]}
  >
    8
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="9"
    onClick={[Function]}
  >
    9
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="10"
    onClick={[Function]}
  >
    10
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="11"
    onClick={[Function]}
  >
    11
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="12"
    onClick={[Function]}
  >
    12
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="13"
    onClick={[Function]}
  >
    13
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="14"
    onClick={[Function]}
  >
    14
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="15"
    onClick={[Function]}
  >
    15
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="16"
    onClick={[Function]}
  >
    16
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="17"
    onClick={[Function]}
  >
    17
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="18"
    onClick={[Function]}
  >
    18
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="19"
    onClick={[Function]}
  >
    19
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="20"
    onClick={[Function]}
  >
    20
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="21"
    onClick={[Function]}
  >
    21
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="22"
    onClick={[Function]}
  >
    22
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="23"
    onClick={[Function]}
  >
    23
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="24"
    onClick={[Function]}
  >
    24
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="25"
    onClick={[Function]}
  >
    25
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="26"
    onClick={[Function]}
  >
    26
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="27"
    onClick={[Function]}
  >
    27
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="28"
    onClick={[Function]}
  >
    28
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="29"
    onClick={[Function]}
  >
    29
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="30"
    onClick={[Function]}
  >
    30
  </div>
  <div
    className="calendar__day
                        calendar__day--today
                        
                        calendar__day--work-out
                    "
    key="31"
    onClick={[Function]}
  >
    31
  </div>
</div>
`;

exports[`should render Calendar without props 1`] = `
<div
  className="calendar"
>
  <div
    className="calendar__prev-month"
    onClick={[Function]}
  >
    <LeftArrowIcon
      size={24}
    />
  </div>
  <div
    className="calendar__month-year"
  >
    December 1969
  </div>
  <div
    className="calendar__next-month"
    onClick={[Function]}
  >
    <RightArrowIcon
      size={24}
    />
  </div>
  <div
    className="calendar__weekday-short-name"
    key="0"
  >
    S
  </div>
  <div
    className="calendar__weekday-short-name"
    key="1"
  >
    M
  </div>
  <div
    className="calendar__weekday-short-name"
    key="2"
  >
    T
  </div>
  <div
    className="calendar__weekday-short-name"
    key="3"
  >
    W
  </div>
  <div
    className="calendar__weekday-short-name"
    key="4"
  >
    T
  </div>
  <div
    className="calendar__weekday-short-name"
    key="5"
  >
    F
  </div>
  <div
    className="calendar__weekday-short-name"
    key="6"
  >
    S
  </div>
  <div
    className="calendar__day--empty"
    key="0"
  />
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="1"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="2"
    onClick={[Function]}
  >
    2
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="3"
    onClick={[Function]}
  >
    3
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="4"
    onClick={[Function]}
  >
    4
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="5"
    onClick={[Function]}
  >
    5
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="6"
    onClick={[Function]}
  >
    6
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="7"
    onClick={[Function]}
  >
    7
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="8"
    onClick={[Function]}
  >
    8
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="9"
    onClick={[Function]}
  >
    9
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="10"
    onClick={[Function]}
  >
    10
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="11"
    onClick={[Function]}
  >
    11
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="12"
    onClick={[Function]}
  >
    12
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="13"
    onClick={[Function]}
  >
    13
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="14"
    onClick={[Function]}
  >
    14
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="15"
    onClick={[Function]}
  >
    15
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="16"
    onClick={[Function]}
  >
    16
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="17"
    onClick={[Function]}
  >
    17
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="18"
    onClick={[Function]}
  >
    18
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="19"
    onClick={[Function]}
  >
    19
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="20"
    onClick={[Function]}
  >
    20
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="21"
    onClick={[Function]}
  >
    21
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="22"
    onClick={[Function]}
  >
    22
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="23"
    onClick={[Function]}
  >
    23
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="24"
    onClick={[Function]}
  >
    24
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="25"
    onClick={[Function]}
  >
    25
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="26"
    onClick={[Function]}
  >
    26
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="27"
    onClick={[Function]}
  >
    27
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="28"
    onClick={[Function]}
  >
    28
  </div>
  <div
    className="calendar__day
                        
                        
                        calendar__day--work-out
                    "
    key="29"
    onClick={[Function]}
  >
    29
  </div>
  <div
    className="calendar__day
                        
                        
                        
                    "
    key="30"
    onClick={[Function]}
  >
    30
  </div>
  <div
    className="calendar__day
                        calendar__day--today
                        
                        calendar__day--work-out
                    "
    key="31"
    onClick={[Function]}
  >
    31
  </div>
</div>
`;
